### PR TITLE
Return original return value.

### PR DESCRIPTION
**Pull request by vaibhavpandeyvpz**
The `withScope` method is very useful to push local scope to error that happen during specific executions. But what if we rely on return values from the execution? It can be handled in a ugly way as below using current code.

```php
$success = false;

Sentry\withScope(function (Sentry\State\Scope $scope) use ($processor, $job, &$success) {
    $scope->setUser([
        'username' => $job->user->username,
    ]);
    $success = $processor->processJob($job);
});

if ($success) {
    // do whatever
}
```

The goal of this PR to simplify above as below:

```php
$success = Sentry\withScope(function (Sentry\State\Scope $scope) use ($processor, $job) {
    $scope->setUser([
        'username' => $job->user->username,
    ]);
    return $processor->processJob($job);
});
```

IMO this is cleaner and doesn't introduce any backwards incompatibility with existing code as well.

### CODE REVIEW

**Review by Jean85 — CHANGES_REQUESTED** [review]
Using short closures would be a less-ugly way to do something like this, but IMO this feature still has its merits!

To go forward, could you please:
 * rebase on `develop` (@ste93cry already retargeted your MR)
 * fix CS
 * add a test
 * add a changelog entry

**Comment by HazAT** [comment]
@ste93cry We had a quick internal discussion and it's fine ✔️ 
We can have small QoL changes like these if it makes usage of our SDK easier and more convenient

**Comment by vaibhavpandeyvpz** [comment]
@Jean85 Ready for review again.

File: tests/FunctionsTest.php [hunk 1]
```diff
@@ -206,6 +206,15 @@ public function testWithScope(): void
         $this->assertTrue($callbackInvoked);
     }
 
+    public function testWithScopeWithCallback(): void
```
Comment (ste93cry) [inline tests/FunctionsTest.php:209]: It doesn't make sense to pass the variable as reference to the lambda function if you then assign the return value of the callback to it from the outer scope. This test can be merged with the previous one instead and we can simply check the return value

File: tests/State/HubAdapterTest.php [hunk 2]
```diff
@@ -105,6 +105,23 @@ public function testWithScope(): void
         HubAdapter::getInstance()->withScope($callback);
     }
 
+    public function testWithScopeWithReturnValue(): void
```
Comment (ste93cry) [inline tests/State/HubAdapterTest.php:108]: As written above, this test can be squashed with the previous one and you can just check that the value returned by the callback is the one you're expecting

File: tests/State/HubAdapterTest.php [hunk 3]
```diff
@@ -105,6 +105,23 @@ public function testWithScope(): void
...
+    {
+        $callback = static function () {
+            return true;
+        };
+
+        $hub = $this->createMock(HubInterface::class);
+        $hub->expects($this->once())
+            ->method('withScope')
+            ->with($callback)
+            ->willReturn(true);
```
Comment (ste93cry) [inline tests/State/HubAdapterTest.php:118]: It would be wiser to use `willReturnArgument()` so that we won't have to bother changing the argument in a future refactoring if we will decide to change the expected value

**Comment by vaibhavpandeyvpz** [comment]
> There are still commits not belonging to this PR that need to be removed 😬

I don't understand how they came in though. Maybe because I did cut the branch off `master`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,8 +2,14 @@
...
 
-- Update the Guzzle tracing middleware to meet the [expected standard](https://develop.sentry.dev/sdk/features/#http-client-integrations) (#1234)
+## 3.3.4 (2021-11-08)
+
+- Update Guzzle tracing middleware to meet the [expected standard](https://develop.sentry.dev/sdk/features/#http-client-integrations) (#1234)
+- Run the test suite against PHP `8.1` (#1245)
+- Avoid overwriting the error level set by the user on the event when capturing an `ErrorException` exception (#1251)
+- Allow installing the project alongside Symfony `6.x` components (#1257)
 - Add `toArray` public method in `PayloadSerializer` to be able to re-use Event serialization
+- The `withScope` methods now return the callback's return value (#1263)
```
Comment (ste93cry) [inline CHANGELOG.md]: This change should go under the `Unreleased` section

**Comment by ste93cry** [comment]
It happened because the branch was originally branched from `master`, which contained more commits than `develop`. When I changed the target branch, the commits remained. I've now merged `master` into `develop`, but a rebase is still needed

File: src/functions.php [hunk 5]
```diff
@@ -94,10 +94,12 @@ function configureScope(callable $callback): void
  * is automatically removed once the operation finishes or throws.
  *
  * @param callable $callback The callback to be executed
+ *
+ * @return mixed|void The callback's return value, upon successful execution
```
Comment (ste93cry) [inline src/functions.php:98]: Same as above

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,14 +2,15 @@
 
 ## Unreleased
 
-- Update the Guzzle tracing middleware to meet the [expected standard](https://develop.sentry.dev/sdk/features/#http-client-integrations) (#1234)
+- Update Guzzle tracing middleware to meet the [expected standard](https://develop.sentry.dev/sdk/features/#http-client-integrations) (#1234)
 - Add `toArray` public method in `PayloadSerializer` to be able to re-use Event serialization
+- The `withScope` methods now return the callback's return value (#1263)
```
Comment (ste93cry) [inline CHANGELOG.md:7]: This change should go under the `Unreleased` section

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,14 +2,15 @@
 
 ## Unreleased
 
-- Update the Guzzle tracing middleware to meet the [expected standard](https://develop.sentry.dev/sdk/features/#http-client-integrations) (#1234)
```
Comment (ste93cry) [inline CHANGELOG.md:5]: 🙏 revert all unnecessary changes to this file

File: src/State/HubInterface.php [hunk 8]
```diff
@@ -54,8 +54,10 @@ public function popScope(): bool;
      * is automatically removed once the operation finishes or throws.
      *
      * @param callable $callback The callback to be executed
+     *
+     * @return mixed|void The callback's return value, upon successful execution
      */
-    public function withScope(callable $callback): void;
+    public function withScope(callable $callback);
```
Comment (ste93cry) [inline src/State/HubInterface.php:60]: Please add PHPStan generics to tie the signature of the callable to the return value of this function:

```
@phpstan-template T

@phpstan-param callable(Scope): T $callback

@phpstan-return T
```

File: tests/State/HubTest.php [hunk 9]
```diff
@@ -167,6 +167,35 @@ public function testWithScope(): void
         $this->assertTrue($callbackInvoked);
     }
 
+    public function testWithScopeWithReturnValue(): void
```
Comment (ste93cry) [inline tests/State/HubTest.php:170]: This test can be merged with the previous one, as suggested for the other similar places

**Review by ste93cry — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (vaibhavpandeyvpz) [reply CHANGELOG.md]: I had moved this already.

**Review by ste93cry — CHANGES_REQUESTED** [review]
Just some more quick changes and we should be good to go! Thanks for the patience

**Comment by ste93cry** [comment]
@vaibhavpandeyvpz are you still interested in completing the work?

**Comment by ste93cry** [comment]
Closing as superseded by @HypeMC's work
